File: server/ruby/app/controllers/hosted_checkout_controller.rb
require_relative '../hosted_checkout/hosted_checkout_mapper'
require_relative '../hosted_checkout/hosted_checkout_service'
require_relative '../payment/payment_details_service'
require_relative '../hosted_checkout/create_hosted_checkout_basic_dto'
require_relative '../configuration/merchant_client_config'
require_relative '../services/payment_utils'
require 'onlinepayments/sdk'

class HostedCheckoutController < ApplicationController

  def initialize_hosted_checkout
    dto = HostedCheckoutMapper.to_empty_dto
    render json: dto.to_h
  end

  def create_hosted_checkout
    create_hosted_checkout_basic_dto = CreateHostedCheckoutBasicDto.new(
      input_params[:amount],
      input_params[:currency],
      input_params[:redirectUrl]
    )

    create_hosted_checkout_request = HostedCheckoutMapper.to_create_hosted_checkout_request(create_hosted_checkout_basic_dto)
    create_hosted_checkout_response = HostedCheckoutService.create_hosted_checkout(create_hosted_checkout_request)
    render json: PaymentUtils.to_camel_case_keys(create_hosted_checkout_response.to_h)
  end

  def outcome
    hosted_checkout_id = params[:hostedCheckoutId]

    payment_details_response = PaymentDetailsService.get_payment_details_for_hosted_checkout(hosted_checkout_id)
    render json: payment_details_response
  end

  private

  def input_params
    params.permit(:amount, :currency, :redirectUrl)
  end

end

File: server/python/hosted_checkout/hosted_checkout_controller.py
from flask import request # type: ignore
from hosted_checkout import hosted_checkout_mapper, hosted_checkout_service
from payment import payment_details_service

def initialize_hosted_checkout():
    return hosted_checkout_mapper.to_empty_dto()

def create_hosted_checkout(hosted_checkout_request_dto):
    hosted_checkout_request = hosted_checkout_mapper.to_hosted_checkout_request(hosted_checkout_request_dto)
    hosted_checkout_response = hosted_checkout_service.create_hosted_checkout(hosted_checkout_request)
    return hosted_checkout_response.to_dictionary()

def outcome():
    hostedCheckoutId = request.args.get('hostedCheckoutId')
    payment_details_response = payment_details_service.get_payment_details_for_hosted_checkout(hostedCheckoutId)
    return payment_details_response.to_dictionary()